src/app/add-item/add-item.component.ts

```diff
@@ -10,59 +10,93 @@ import { first } from 'rxjs/operators';
+  dialogTitle: string;
+  dialogActionButton: string;
+  projects?: Project[];
+  isSavingDialogData: boolean = false;
+  id: string;
+  loaderVisible: boolean = true;
 
   constructor(
-    public dialogRef: MatDialogRef<AddItemComponent>,
+    public dialogRef: MatDialogRef<AddOrUpdateActionItemComponent>,
     private formBuilder: FormBuilder,
     private dialogProjects: ProjectsService,
-    private tasksService: TasksService
+    private tasksService: TasksService,
+    @Inject(MAT_DIALOG_DATA) public data: any
   ) {}
+
   ngOnInit() {
+    if (this.data.createDialog === true) {
+      this.dialogTitle = 'Create Action Item';
+      this.dialogActionButton = 'Create';
+    } else {
+      this.dialogTitle = 'Edit Action Item';
+      this.dialogActionButton = 'Edit';
+    }
     this.dialogProjects
       .getProjectsNames()
       .pipe(first())
-      .subscribe(projects => (this.projects = projects));
+      .subscribe(projects => {
+        this.projects = projects;
+        this.loaderVisible = false;
+      });
     this.createForm();
+    this.dialogForm.patchValue({
+      title: this.data.item ? this.data.item.title : '',
+      projectName: this.data.item ? this.data.item.projectName : '',
+      dueDate: this.data.item ? this.data.item.dueDate : '',
+      description: this.data.item ? this.data.item.description : '',
+      id: this.data.item ? this.data.item.id : ''
+    });
+    this.id = this.data.id;
+    return this.dialogForm.value;
   }
   private createForm(): void {
     this.dialogForm = this.formBuilder.group({
-      name: ['', Validators.required],
-      project: ['', Validators.required],
+      title: ['', Validators.required],
+      projectName: ['', Validators.required],
       dueDate: '',
-      description: ''
+      description: '',
+      id: ''
     });
   }
+  saveForm(): void {
+    this.isSavingDialogData = true;
+    this.id ? this.editItem() : this.createActionItem();
+  }
   createActionItem(): void {
     const newItem = this.formNewActionItem();
-    this.isCreatingActionItem = true;
     this.dialogForm.disable();
     this.tasksService
       .add(newItem)
       .pipe(first())
       .subscribe(actionItem => {
         this.dialogRef.close(actionItem);
-        this.isCreatingActionItem = false;
       });
   }
   private formNewActionItem(): AddActionItem {
-    const title = this.dialogForm.get('name');
-    const projectName = this.dialogForm.get('project');
+    const title = this.dialogForm.get('title');
+    const projectName = this.dialogForm.get('projectName');
     const dueDate = this.dialogForm.get('dueDate');
     if (!title || !projectName || !dueDate) {
       throw 'Invalid Action Item data';
     }
     const newItem: AddActionItem = {
       title: title.value,
-      projectName: projectName.value.name,
+      projectName: projectName.value,
       type: 'General',
       completed: '0',
       dueDate: dueDate.value
     };
     return newItem;
   }
+  editItem(): void {
```

Comment:
**SUGGGESTION**: It's generally good to have a convention of the naming - `createActionItem` vs `editItem`. We should name it either all with or without Action word.

Reply:
still not applied